packages/flutter/lib/src/material/bottom_sheet_scroll_controller.dart

```diff
@@ -0,0 +1,478 @@
+// Copyright 2018 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:math' as math;
+
+import 'package:flutter/gestures.dart';
+import 'package:flutter/physics.dart';
+import 'package:flutter/scheduler.dart';
+import 'package:flutter/widgets.dart';
+
+import 'debug.dart';
+import 'scaffold.dart';
+
+const double _kBottomSheetMinHeight = 56.0;
+
+/// Controls a scrollable widget that is not fully visible on screen yet. While
+/// the [top] value is between [minTop] and [maxTop], scroll events will drive
+/// [top]. Once it has reached [minTop] or [maxTop], scroll events will drive
+/// [offset]. The [top] value is guaranteed not to be clamped between
+/// [minTop] and [maxTop].
+///
+/// This controller would typically be created and listened to by a parent
+/// widget such as a [Positioned] or an [Align], and then either passed in
+/// directly or used as a [PrimaryScrollController] by a [Scrollable] descendant
+/// of that parent.
+///
+/// See also:
+///
+///  * [BottomSheetScrollPosition], which manages the positioning logic for
+///    this controller.
+///  * [PrimaryScrollController], which can be used to establish a
+///    [BottomSheetScrollController] as the primary controller for
+///    descendants.
+class BottomSheetScrollController extends ScrollController {
+  /// Creates a new [BottomSheetScrollController].
+  ///
+  /// The [top] and [minTop] parameters must not be null. If [maxTop]
+  /// is provided as null, it will be defaulted to the
+  /// [MediaQueryData.size.height].
+  BottomSheetScrollController({
+    double initialScrollOffset = 0.0,
+    double initialHeightPercentage = 0.5,
+    this.minTop = 0.0,
+    String debugLabel,
+    this.isPersistent = false,
+    @required BuildContext context,
+    bool forFullScreen = false,
+  })  : assert(initialHeightPercentage != null),
+        assert(context != null),
+        assert(minTop != null),
+        assert(isPersistent != null),
+        _initialTop = _topFromInitialHeightPercentage(initialHeightPercentage, context, forFullScreen),
+        maxTop = _calculateMaxTop(initialHeightPercentage, context, isPersistent),
+        super(
+          debugLabel: debugLabel,
+          initialScrollOffset: initialScrollOffset,
+      ) {
+    // If the BottomSheet's child doesn't have a Scrollable widget in it that
+    // inherits our PrimaryScrollController, it will never become visible.
+    assert(() {
+      SchedulerBinding.instance.addPostFrameCallback((Duration duration) {
```

Comment:
The problem I'm trying to solve is that if you use this scroll controller on a bottom sheet like this:

```
BottomSheet(
  child: ListView(children: ...), // not setting primary: true
  ...
),
```

it won't animate onto the screen.  But you're right, this approach would mean if someone created the BottomSheetScrollController ahead of time for some reason and later used it on a BottomSheet a few frames latter, this assert would fire.  I suppose it could be turned into a warning instead of an assert in that case, but I'm not sure of the best way to handle that and still not confuse a developer who can't see their bottom sheet.